flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/source/split/BigQuerySourceSplitAssigner.java

```diff
@@ -0,0 +1,249 @@
+
+import org.apache.flink.annotation.Internal;
+
+import org.apache.flink.shaded.guava30.com.google.common.collect.Lists;
+
+import com.google.cloud.bigquery.storage.v1.CreateReadSessionRequest;
+import com.google.cloud.bigquery.storage.v1.DataFormat;
+import com.google.cloud.bigquery.storage.v1.ReadSession;
+import com.google.cloud.bigquery.storage.v1.ReadSession.TableModifiers;
+import com.google.cloud.bigquery.storage.v1.ReadSession.TableReadOptions;
+import com.google.cloud.flink.bigquery.common.config.BigQueryConnectOptions;
+import com.google.cloud.flink.bigquery.services.BigQueryServices;
+import com.google.cloud.flink.bigquery.services.BigQueryServicesFactory;
+import com.google.cloud.flink.bigquery.services.QueryResultInfo;
+import com.google.cloud.flink.bigquery.source.config.BigQueryReadOptions;
+import com.google.cloud.flink.bigquery.source.enumerator.BigQuerySourceEnumState;
+import com.google.protobuf.Timestamp;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** A simple split assigner based on the BigQuery {@link ReadSession} streams. */
+@Internal
+public class BigQuerySourceSplitAssigner {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BigQuerySourceSplitAssigner.class);
+
+    private final BigQueryReadOptions readOptions;
+
+    private final ArrayDeque<String> remainingTableStreams;
+    private final List<String> alreadyProcessedTableStreams;
+    private final ArrayDeque<BigQuerySourceSplit> remainingSourceSplits;
+    private final Map<String, BigQuerySourceSplit> assignedSourceSplits;
+    private boolean initialized;
+
+    public BigQuerySourceSplitAssigner(
+            BigQueryReadOptions readOptions, BigQuerySourceEnumState sourceEnumState) {
+        this.readOptions = readOptions;
+        this.remainingTableStreams = new ArrayDeque<>(sourceEnumState.getRemaniningTableStreams());
+        this.alreadyProcessedTableStreams = sourceEnumState.getCompletedTableStreams();
+        this.remainingSourceSplits = new ArrayDeque<>(sourceEnumState.getRemainingSourceSplits());
+        this.assignedSourceSplits = sourceEnumState.getAssignedSourceSplits();
+        this.initialized = sourceEnumState.isInitialized();
+    }
+
+    /**
+     * Reviews the read options argument and see if a query has been configured, in that case run
+     * that query and then return a modified version of the connect options pointing to the
+     * temporary location (project, dataset and table) of the query results.
+     *
+     * @return The BigQuery connect options with the right project, dataset and table given the
+     *     specified configuration.
+     */
+    BigQueryConnectOptions checkOptionsAndRunQueryIfNeededReturningModifiedOptions() {
```

Comment:
This function has too many responsibilities.
We can make multiple functions: 
`runQuery`: runs the specified query and returns result,
`fetchOptionsFromQueryRun` can call `runQuery` and generate options from the result, and returns an `Optional.empty` if query is not provided / empty.
The caller can choose the fallback behaviour i.e. in this case the caller code will be:
`getOptionsFromQuery().orElse(this.readOptions.getBigQueryConnectOptions());`